hadoop-hdds/docs/content/design/overwrite-key-only-if-unchanged.md

```diff
@@ -0,0 +1,190 @@
+
+Multi-part keys need more investigation and hence are also excluded in the initial version.
+
+## Changes Required
+
+In order to enable the above steps on Ozone, several small changes are needed.
+
+### Wire Protocol
+
+1. The expectedGeneration needs to be added to the KeyInfo protobuf object so it can be stored in the openKey table.
+2. The expectedGeneration needs to be added to the keyArgs protobuf object, which is passed from the client to OM when creating a key.
+
+No new messages need to be defined.
+
+### On OM
+
+No new OM handlers are needed. The existing OpenKey and CommitKey handlers will receive the new expectedGeneration and perform the checks.
+
+No new locks are needed on OM. As part of the openKey and commitKey, there are existing locks taken to ensure the key open / commit is atomic. The new checks are performed under those locks, and come down to a couple of long comparisons, so add negligible overhead.
+
+### On The Client
+
+ 1. We need to allow the updateID (called generation on the client) of an existing key to be accessible when an existing details are read, by adding it to OzoneKey and OzoneKeyDetails. There are internal object changes and do no impact any APIs.
+ 2. To pass the expectedGeneration to OM on key open, it would be possible to overload the existing OzoneBucket.createKey() method, which already has several overloaded versions, or create a new explicit method on Ozone bucket called rewriteKey, passing the expectedGeneration, eg:
+ 
+ ```
+
+ public OzoneOutputStream rewriteKey(String volumeName, String bucketName, String keyName, long size, long expectedGeneration, ReplicationConfig replicationConfigOfNewKey)
+      throws IOException 
+      
+// Can also add an overloaded version of these methods to pass a metadata map, as with the existing
+// create key method.      
+
+	  
+ ```
+This specification is roughly in line with the exiting createKey method:
+
+```
+  public OzoneOutputStream createKey(
+      String volumeName, String bucketName, String keyName, long size,
+      ReplicationConfig replicationConfig,
+      Map<String, String> metadata)
+```
+
+An alternative, is to create a new overloaded createKey, but it is probably less confusing to have the new rewriteKey method:
+
+```
+  public OzoneOutputStream createKey(
+      String volumeName, String bucketName, String keyName, long size,
+      ReplicationConfig replicationConfig, long expectedUpdateID)
+```
+
+The intended usage of this API, is that the existing key details are read, then used to open the new key, and then data is written, eg:
+
+```
+OzoneKeyDetails exisitingKey = bucket.getKey(keyName);
+try (OutputStream os = bucket.rewriteKey(existingKey.getBucket, existingKey.getVolume, 
+    existingKey.getKeyName, existingKey.getSize(), existingKey.getGeneration(), newRepConfig) {
+  os.write(bucket.readKey(keyName))
```

Comment:
```
ostream rewriteKey(key, repType) {
    genID = getInfo(key)
    return create(key, genID)
}
```

Reply:
The key is returned to the client as an object (KeyInfo) - in an earlier version I simply passed the keyInfo, but you and @kerneltime didn't seem to like that, so I changed it to look like the existing create API.

In the general case whatever application is using the rewrite API has to pull a keys details and then decided based on the key metadata or content if it wants to rewrite it or not. Therefore having the rewrite method pull the key details will result in:

1. Two pulls from OM to get the key info when one would have done.
2. The potential for different key details to be returned between 1 and 2, and the details from 2 may not want to be overwritten by the application.

The point of the API is that the application pulls some details and then makes a decision based on those - this key needs updated. And it determines that "this key" has not changed by the generation / updateID it received then.

Therefore I believe the method must be passed the keyInfo or the details of the key it is to overwrite.

Reply:
> Therefore I believe the method must be passed the keyInfo or the details of the key it is to overwrite.

Ok I see my comment here wasn't clear. I was trying to abstract generation ID inside the rewrite method, not get rid of the key parameters. Key parameters here are fine.

> in an earlier version I simply passed the keyInfo, but you and @kerneltime didn't seem to like that

This makes the objection sound arbitrary. There was technical validation to this point. Separating the key parameters like this is consistent with `RpcClient#createKey`. This also leaves `OMKeyInfo` as an output given to the client like its current usage, not input from the client to OM. That's what `KeyArgs` would be for.

> The point of the API is that the application pulls some details and then makes a decision based on those - this key needs updated.

This actually relates to both this comment and [this one](https://github.com/apache/ozone/pull/6482#issuecomment-2075163977)

> This does not work in the general case, where a client reads a key, inspects it and decides that it needs rewritten

I think I see the confusion. They *do* work for the use case presented here in the document, where there are no reads between when the generation ID is read and the rewrite starts. "Use Cases" should really be its own section in the doc with more thorough examples, as it helps answer questions like this.
Where the suggestion here *doesn't* work is for a use case that your comments seem to imply but the document does not define:

```
OzoneKeyDetails exisitingKey = bucket.getKey(keyName);
if (existingKey.getReplicationType() == RATIS) { // Important addition that changes the guarantees required from the client methods and API
  try (OutputStream os = bucket.rewriteKey(existingKey.getBucket, existingKey.getVolume, 
      existingKey.getKeyName, existingKey.getSize(), existingKey.getGeneration(), EC) {
    os.write(bucket.readKey(keyName))
  }
}
```

So the proposal here looks good, but a use cases section will help illustrate both now and in the future why certain decisions were made instead of others.